### ISSUE TITLE
classify detected object vs combined in one step

### ISSUE THREAD

**Issue by JasonRuan5** [AlexeyAB/darknet#3086]
Hi @AlexeyAB, Thanks for the great darknet framework.  After read issue #2552, I’ve 3 related questions:
1.	At #2552, you suggested to train classifier first, then train detector with extracted pre-trained weights.  I just wonder if the final weights (assuming the pre-trained weights will be updated and combined) will be as good as the pre-trained weights to identify the cells correctly?  Would be better to first detect cells, then classify the detected cells?
2.	There is a valid file specified in classification.data, but I could find where the valid files are used during train.  Do you know where is used?
3.	What classification model do you recommend to train for cell classification?  I’ve tried with darknet.cfg and looks pretty good.  Since all images to be trained are small size, e.g. 64x64, is it better to change the network size to 64x64, too?

**Comment by AlexeyAB** [comment]
@JasonRuan5 Hi,

> At #2552, you suggested to train classifier first, then train detector with extracted pre-trained weights. I just wonder if the final weights (assuming the pre-trained weights will be updated and combined) will be as good as the pre-trained weights to identify the cells correctly? 

What do you mean?

> Would be better to first detect cells, then classify the detected cells?

It can give better accuracy bot lower speed.

> There is a valid file specified in classification.data, but I could find where the valid files are used during train. Do you know where is used?

Are you about this line? https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/cfg/imagenet1k.data#L3
It is used to `topK` calculation - top1 and top5 error (%).

> What classification model do you recommend to train for cell classification? I’ve tried with darknet.cfg and looks pretty good. Since all images to be trained are small size, e.g. 64x64, is it better to change the network size to 64x64, too?

Yes, its a good idea to use `darknet53.cfg` with `width=64 height=64`

**Comment by JasonRuan5** [comment]
Hi @AlexeyAB,

Thanks so much for such a quick response.  Yes, I meant valid  = /data/imagenet/imagenet1k.valid.list and great I see now (it is used in topk calculation).  Do you think inception model would be a good classifier for cell?  If so, does darknet support inception?  Thanks once again.

**Comment by AlexeyAB** [comment]
@JasonRuan5 

Darknet currently doesn't support Inception models (doesn't support non-sqaure kernels).
Also may be it makes sense to use smaller model for such small network size (64x64).
You can use one of these models: https://pjreddie.com/darknet/imagenet/

**Comment by JasonRuan5** [comment]
Hi @AlexeyAB,

I’ve tried with several models, e.g. cifar_small, resenet50, vgg-16, and darknet.  The darknet seems to give the smallest loss, but the top1 accuracy varies a lot – oscillates between 0.4 and 0.75.  I am still new to deep learning, so I only changed the network width and height to 64, and the last convolution layer filters to 4.  We have about 10,000 of 64x64 images for training and 1000 images for validation and only need to classify 4 classes.  Not like apple and orange, the 4 type of cells looks similar, the main differences are nuclei size and morphology. I’ve also tried data augmentation by setting:
angle=15
hue = .1
saturation=.75
exposure=.75
It indeed helps -- the variation becomes between 0.6 and 0.8
My questions are:
1.	What could cause the large variations?  
2.	Any parameters that we could modify?
3.	Should we do random crop?
Thanks!

**Comment by JasonRuan5** [comment]
Actually, when I monitored longer, the data augmentation seems not help much, the variation is still large.  The train loss is is around 0.15 and accuracy oscillates between 0.5 and 0.75.

**Comment by AlexeyAB** [comment]
> Actually, when I monitored longer, the data augmentation seems not help much, the variation is still large. The train loss is is around 0.15 and accuracy oscillates between 0.5 and 0.75.

Can you show Loss & TopK chart?

----

> What could cause the large variations?

If Training and Validation set are very different. Did you divide inital dataset to Training and Validation randomly?

> Any parameters that we could modify?

You can try to use 
```
angle=180
hue = .3`
saturation=1.5
exposure=1.5
```

> Should we do random crop?

Yes. The range depends on how much Validation/Test dataset differs from Training dataset.
```
min_crop=32
max_crop=96
```

* Can you show examples of Training and Validation images?
* Can you attach your cfg-file?

**Comment by JasonRuan5** [comment]
@AlexeyAB 
Yes, we have divided initial dataset to Training and Validation randomly.  You can view or download loss-topk chart along with the examples of training and validation images from the link below.
http://www.vigenetech.com/plugin/content.html
We’ve put 3 different size for the train and validation images.  Do you think we should use size 64, 96, or 128? You can also view the cfg-file below:
http://www.vigenetech.com/plugin/darknet_64.txt
Thanks so much for your kind help!

**Comment by JasonRuan5** [comment]
@AlexeyAB 
I've trained again based on your suggestion, the results are indeed much improved as you can see in attachment.  The fluctuation comes down within 6.5 and 7.5.

![loss_topk](https://user-images.githubusercontent.com/31511478/57531009-83381700-7306-11e9-8825-2afddd8f875d.JPG)

It would be very hard for me to figure it out why the new cfg is much better than the old one.  May I ask:

1. What could be the main cause of the large fluctuation from the previous cfg-file and what parameters made the new cfg much better?
2. What else can I change to make even better?
3. If we use size 96 or 128 instead of 64 to train, will we get better topk?

Thanks again!

**Comment by AlexeyAB** [comment]
@JasonRuan5 

> What could be the main cause of the large fluctuation from the previous cfg-file and what parameters made the new cfg much better?

Attach both cfg-files


> If we use size 96 or 128 instead of 64 to train, will we get better topk?

Yes only if image sizes more than 64.

**Comment by JasonRuan5** [comment]
@AlexeyAB 

Here are two cfg-files: darknet_64_old is the one I used and darknet_64_new is modified based on your advice a week ago.  I was surprised to see that much improvement, but would be great to understand why (thanks again).
 
[darknet_64_old.txt](https://github.com/AlexeyAB/darknet/files/3169185/darknet_64_old.txt)
[darknet_64_new.txt](https://github.com/AlexeyAB/darknet/files/3169186/darknet_64_new.txt)

Good, we will try size 128 and hope we can get even better topk!

**Comment by AlexeyAB** [comment]
@JasonRuan5 

> What could be the main cause of the large fluctuation from the previous cfg-file and what parameters made the new cfg much better?


You used more suitable crop params for your width= and height= in new-cfg file.

Also `hue = .3` makes your model more independent to color, due to more aggressive color-data-augmentation.

-----

> Good, we will try size 128 and hope we can get even better topk!

Use
```
height=128
width=128
min_crop=96
max_crop=160
```

And use `hue = .3` instead of `hue = .3'`

**Comment by JasonRuan5** [comment]
@AlexeyAB 

Great and thanks! 

When train on size 128, should we use the weights from size 64 as pre-train weights or better to start from scratch?

**Comment by AlexeyAB** [comment]
If you have enough time then better to train from scratch.

**Comment by JasonRuan5** [comment]
@AlexeyAB 
I've trained and test with size 128.  It is indeed better as shown in attached.  Thanks!!! 
![loss_topk_128](https://user-images.githubusercontent.com/31511478/57667552-f923d300-75d1-11e9-88b3-39af5f2a70ea.png)
